modules/xphoto/test/test_grayworld.cpp

```diff
@@ -0,0 +1,90 @@
+#include "test_precomp.hpp"
+
+namespace cvtest {
+
+    using namespace cv;
+
+    void ref_autowbGrayworld(InputArray _src, OutputArray _dst, float thresh)
+    {
+        Mat src = _src.getMat();
+
+        _dst.create(src.size(), src.type());
+        Mat dst = _dst.getMat();
+
+        int width  = src.cols,
+            height = src.rows,
+            N      = width*height,
+            N3     = N*3;
+
+        // Calculate sum of pixel values of each channel
+        const uchar* src_data = src.ptr<uchar>(0);
+        unsigned long sum1 = 0, sum2 = 0, sum3 = 0;
+        int i = 0;
+        unsigned int minRGB, maxRGB, thresh255 = round(thresh * 255);
+        for ( ; i < N3; i += 3 )
+        {
+            minRGB = std::min(src_data[i], std::min(src_data[i + 1], src_data[i + 2]));
+            maxRGB = std::max(src_data[i], std::max(src_data[i + 1], src_data[i + 2]));
+            if ( (maxRGB - minRGB) * 255 > thresh255 * maxRGB ) continue;
+            sum1 += src_data[i];
+            sum2 += src_data[i + 1];
+            sum3 += src_data[i + 2];
+        }
+
+        // Find inverse of averages
+        double inv1 = sum1 == 0 ? 0.f : (double)N / (double)sum1,
+               inv2 = sum2 == 0 ? 0.f : (double)N / (double)sum2,
+               inv3 = sum3 == 0 ? 0.f : (double)N / (double)sum3;
+
+        // Find maximum
+        double inv_max = std::max(std::max(inv1, inv2), inv3);
+
+        // Scale by maximum
+        if ( inv_max > 0 )
+        {
+            inv1 = (double) inv1 / inv_max;
+            inv2 = (double) inv2 / inv_max;
+            inv3 = (double) inv3 / inv_max;
+        }
+
+        // Fixed point arithmetic, mul by 2^8 then shift back 8 bits
+        unsigned int i_inv1 = inv1 * (1 << 8),
```

Comment:
use "int" instead of "unsigned int" and use cvRound to convert double's to int's.